# Import Libraries

In [1]:
import numpy as np
import pandas as pd
pd.set_option("max_columns", None)
from scipy.stats import chi2_contingency
from sklearn.linear_model import Lasso
from sklearn.linear_model import LassoCV
from sklearn.linear_model import Ridge
from sklearn.model_selection import train_test_split
from sklearn.metrics import r2_score
from sklearn.model_selection import StratifiedKFold

from keras.layers.normalization import BatchNormalization
from keras.layers.core import Activation, Dropout
from keras.optimizers import Adadelta
from keras.layers.advanced_activations import LeakyReLU
from keras.wrappers.scikit_learn import KerasRegressor
from sklearn.model_selection import KFold
from sklearn.model_selection import cross_val_score
from keras.models import Sequential
from keras.layers import Dense

Using TensorFlow backend.


# Import Data

In [2]:
data = pd.read_csv('fashion_data_on_Instagram.csv')
data.set_index('UserId', inplace = True)
data.head()

Followings  Followers  MediaCount  \
UserId                                                              
1171579752855683619_212070047        518.0     9840.0         541   
1171594777274371222_176762322       7333.0     2300.0         272   
1171407552643586413_581125501        131.0      605.0         106   
1171407552643586413_581125501        131.0      605.0         106   
1171508187966229230_2797323089       132.0      456.0          56   

                                  BrandName BrandCategory  \
UserId                                                      
1171579752855683619_212070047   abercrombie   High street   
1171594777274371222_176762322   abercrombie   High street   
1171407552643586413_581125501   abercrombie   High street   
1171407552643586413_581125501   abercrombie   High street   
1171508187966229230_2797323089  abercrombie   High street   

                                                                         Hashtags  \
UserId                                                                              
1171579752855683619_212070047   beautiful, summer, fashion, love, cute, food, ...   
1171594777274371222_176762322   teen, model, brunette, selfie, hollister, snap...   
1171407552643586413_581125501   alexandani, cute, llbean, beanboots, anthropol...   
1171407552643586413_581125501   alexandani, cute, llbean, beanboots, anthropol...   
1171508187966229230_2797323089  abercrombieandfitch, shopping, love, hollister...   

                                                                          Caption  \
UserId                                                                              
1171579752855683619_212070047     We were born to be REAL, not to be perfect. ...   
1171594777274371222_176762322     Gotta run, but first, let me take selfie. Me...   
1171407552643586413_581125501                         dress to impress\U0001f457    
1171407552643586413_581125501                         dress to impress\U0001f457    
1171508187966229230_2797323089    #love #shopping #shoppen #hollister #abercro...   

                                                                           ImgURL  \
UserId                                                                              
1171579752855683619_212070047   https://scontent.cdninstagram.com/t51.2885-15/...   
1171594777274371222_176762322   https://scontent.cdninstagram.com/t51.2885-15/...   
1171407552643586413_581125501   https://scontent.cdninstagram.com/t51.2885-15/...   
1171407552643586413_581125501   https://scontent.cdninstagram.com/t51.2885-15/...   
1171508187966229230_2797323089  https://scontent.cdninstagram.com/t51.2885-15/...   

                                Likes  Comments   CreationTime  \
UserId                                                           
1171579752855683619_212070047      97          0    1453883211   
1171594777274371222_176762322      94          0    1453885002   
1171407552643586413_581125501      91          3    1453862683   
1171407552643586413_581125501      94          3    1453862683   
1171508187966229230_2797323089      9          1    1453874680   

                                                                    Link  \
UserId                                                                     
1171579752855683619_212070047   https://www.instagram.com/p/BBCSZPzAMIj/   
1171594777274371222_176762322   https://www.instagram.com/p/BBCVz4YJEyW/   
1171407552643586413_581125501   https://www.instagram.com/p/BBBrPZ2h9lt/   
1171407552643586413_581125501   https://www.instagram.com/p/BBBrPZ2h9lt/   
1171508187966229230_2797323089  https://www.instagram.com/p/BBCCH1zMEru/   

                                  Selfie  BodySnap  Marketing  ProductOnly  \
UserId                                                                       
1171579752855683619_212070047   0.000885  0.336438   0.741220     0.009454   
1171594777274371222_176762322   0.003912  0.995503   0.001728     0.002125   
1171407552643586413_5811

# Clean Data

### Let's See if there are any NaNs in the data

In [3]:
hashtag = data.copy()
hashtag['Hashtags'] = data['Hashtags'].isna()
hashtag.groupby('Hashtags')['MediaCount'].count()

Hashtags
False    24606
True       146
Name: MediaCount, dtype: int64

In [4]:
Caption = data.copy()
Caption['Caption'] = data['Caption'].isna()
Caption.groupby('Caption')['MediaCount'].count()

Caption
False    24717
True        35
Name: MediaCount, dtype: int64

In [5]:
set(data['Comments '].isna())

{False}

##### So there exist NaN values in Hashtags and Captions. There are no other columns that contain NaN.

In [6]:
data.isna().groupby('Hashtags')['Followings'].count()

Hashtags
False    24606
True       146
Name: Followings, dtype: int64

##### There are 146 NaN rows in hashtags. It is possible that this is simply due to the fact that some posts do not have hashtags.

In [7]:
data.isna().groupby('Caption')['Followings'].count()

Caption
False    24717
True        35
Name: Followings, dtype: int64

##### There are 35 NaN rows in Caption

### Maybe the caption or hashtag information is missing because there is either only a caption or only hashtags....

In [8]:
hashtag_nan = list(data['Hashtags'].isna())
caption_nan = list(data['Caption'].isna())

In [9]:
count = 0
for i in range(len(hashtag_nan)):
    if hashtag_nan[i] and caption_nan[i]:
        count += 1
print('The number of times both hashtag_nan and caption_nan is true is', count)

The number of times both hashtag_nan and caption_nan is true is 18


##### So we have missing hashtag both 18/35 times that we have a missing caption. So it is not exactly correlated. Maybe we can just drop the rows that have missing caption or hashtag if we need to use these, or claim that there was not hashtag or caption for these rows.

In [10]:
chi2_contingency([[hashtag_nan.count(True), hashtag_nan.count(False)], [caption_nan.count(True), caption_nan.count(False)]])

(67.09615038449334, 2.585828482081104e-16, 1, array([[   90.5, 24661.5],
        [   90.5, 24661.5]]))

### I want to try and see if I can make brand category numeric

In [11]:
brand_category = pd.get_dummies(data['BrandCategory'])

In [12]:
brand_category.head()

Designer  High street  Mega couture  \
UserId                                                                
1171579752855683619_212070047          0            1             0   
1171594777274371222_176762322          0            1             0   
1171407552643586413_581125501          0            1             0   
1171407552643586413_581125501          0            1             0   
1171508187966229230_2797323089         0            1             0   

                                Small couture  
UserId                                         
1171579752855683619_212070047               0  
1171594777274371222_176762322               0  
1171407552643586413_581125501               0  
1171407552643586413_581125501               0  
1171508187966229230_2797323089              0

In [13]:
data_new = pd.concat([data, brand_category], sort = False, axis = 1)
data_new.head()

Followings  Followers  MediaCount  \
UserId                                                              
1171579752855683619_212070047        518.0     9840.0         541   
1171594777274371222_176762322       7333.0     2300.0         272   
1171407552643586413_581125501        131.0      605.0         106   
1171407552643586413_581125501        131.0      605.0         106   
1171508187966229230_2797323089       132.0      456.0          56   

                                  BrandName BrandCategory  \
UserId                                                      
1171579752855683619_212070047   abercrombie   High street   
1171594777274371222_176762322   abercrombie   High street   
1171407552643586413_581125501   abercrombie   High street   
1171407552643586413_581125501   abercrombie   High street   
1171508187966229230_2797323089  abercrombie   High street   

                                                                         Hashtags  \
UserId                                                                              
1171579752855683619_212070047   beautiful, summer, fashion, love, cute, food, ...   
1171594777274371222_176762322   teen, model, brunette, selfie, hollister, snap...   
1171407552643586413_581125501   alexandani, cute, llbean, beanboots, anthropol...   
1171407552643586413_581125501   alexandani, cute, llbean, beanboots, anthropol...   
1171508187966229230_2797323089  abercrombieandfitch, shopping, love, hollister...   

                                                                          Caption  \
UserId                                                                              
1171579752855683619_212070047     We were born to be REAL, not to be perfect. ...   
1171594777274371222_176762322     Gotta run, but first, let me take selfie. Me...   
1171407552643586413_581125501                         dress to impress\U0001f457    
1171407552643586413_581125501                         dress to impress\U0001f457    
1171508187966229230_2797323089    #love #shopping #shoppen #hollister #abercro...   

                                                                           ImgURL  \
UserId                                                                              
1171579752855683619_212070047   https://scontent.cdninstagram.com/t51.2885-15/...   
1171594777274371222_176762322   https://scontent.cdninstagram.com/t51.2885-15/...   
1171407552643586413_581125501   https://scontent.cdninstagram.com/t51.2885-15/...   
1171407552643586413_581125501   https://scontent.cdninstagram.com/t51.2885-15/...   
1171508187966229230_2797323089  https://scontent.cdninstagram.com/t51.2885-15/...   

                                Likes  Comments   CreationTime  \
UserId                                                           
1171579752855683619_212070047      97          0    1453883211   
1171594777274371222_176762322      94          0    1453885002   
1171407552643586413_581125501      91          3    1453862683   
1171407552643586413_581125501      94          3    1453862683   
1171508187966229230_2797323089      9          1    1453874680   

                                                                    Link  \
UserId                                                                     
1171579752855683619_212070047   https://www.instagram.com/p/BBCSZPzAMIj/   
1171594777274371222_176762322   https://www.instagram.com/p/BBCVz4YJEyW/   
1171407552643586413_581125501   https://www.instagram.com/p/BBBrPZ2h9lt/   
1171407552643586413_581125501   https://www.instagram.com/p/BBBrPZ2h9lt/   
1171508187966229230_2797323089  https://www.instagram.com/p/BBCCH1zMEru/   

                                  Selfie  BodySnap  Marketing  ProductOnly  \
UserId                                                                       
1171579752855683619_212070047   0.000885  0.336438   0.741220     0.009454   
1171594777274371222_176762322   0.003912  0.995503   0.001728     0.002125   
1171407552643586413_5811

##### Now the brand category is represented in the last four columns of the dataframe with one_hot_encoding.

### In order to model with these we need to normalize the variables

# Model with Numeric Data Only

First, lets try to build a model with only the numeric data. This will be easy to do and will allow us to understand whether any of these numeric factors influence the number of likes/comments a user gets. 

### Likes Model

In [18]:
X = np.array(data_new[['Selfie', 'BodySnap', 'Marketing', 'ProductOnly', 'NonFashion',
       'Face', 'Logo', 'BrandLogo', 'Smile', 'Outdoor', 'NumberOfPeople',
       'NumberOfFashionProduct', 'Anger', 'Contempt', 'Disgust', 'Fear',
       'Happiness', 'Neutral', 'Sadness', 'Surprise', 'Designer',
       'High street', 'Mega couture', 'Small couture']])
y = np.array(data_new['Likes'])

In [41]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.33, random_state=42)

In [19]:
def standardize(train, test):


    mean = np.mean(train, axis=0)
    std = np.std(train, axis=0)+0.000001

    X_train = (train - mean) / std
    X_test = (test - mean) /std
    return X_train, X_test

In [43]:
X_train, X_test = standardize(X_train, X_test)

In [44]:
epsilon = 1e-200
y_train = np.log(y_train + epsilon)
y_test = np.log(y_test + epsilon)

Let's use a Lasso Regression model to regress based on the likes that we receive. I want to use lasso because lasso will send the unimportant variables to zero allowing us to see which features are important. 

In [27]:
lasso = Lasso(alpha = 0.1, max_iter = 10000).fit(X_train, y_train)

In [28]:
predictions = lasso.predict((X_test))
r2_score(y_test, predictions)

0.02503670198698471

Maybe ridge regression will work better?

In [29]:
ridge = Ridge(alpha = 0.1).fit(X_train, y_train)

In [30]:
predictions = ridge.predict((X_test))
r2_score(y_test, predictions)

0.026527240805967134

### Try with stratified crossvalidation

In [23]:
cv = StratifiedKFold(5)
lasso = Lasso(alpha = 0.1, max_iter = 10000)
for train, test in cv.split(X, y):
    X_train, X_test = standardize(X[train], X[test])
    epsilon = 1e-200
    y_train = np.log(y[train] + epsilon)
    y_test = np.log(y[test] + epsilon)
    
    lasso.fit(X_train, y_train)
    predictions = lasso.predict((X_test))
    print(r2_score(y_test, predictions))

/Users/simi/opt/anaconda3/lib/python3.7/site-packages/sklearn/model_selection/_split.py:657: Warning: The least populated class in y has only 1 members, which is too few. The minimum number of members in any class cannot be less than n_splits=5.
  % (min_groups, self.n_splits)), Warning)


0.03308285132362676
0.01662791735700586
0.02361500128922933
0.006028356774259169
0.030367617948803827


In [24]:
cv = StratifiedKFold(5)
ridge = Ridge(alpha = 0.1)
for train, test in cv.split(X, y):
    X_train, X_test = standardize(X[train], X[test])
    epsilon = 1e-200
    y_train = np.log(y[train] + epsilon)
    y_test = np.log(y[test] + epsilon)
    
    ridge.fit(X_train, y_train)
    predictions = lasso.predict((X_test))
    print(r2_score(y_test, predictions))

/Users/simi/opt/anaconda3/lib/python3.7/site-packages/sklearn/model_selection/_split.py:657: Warning: The least populated class in y has only 1 members, which is too few. The minimum number of members in any class cannot be less than n_splits=5.
  % (min_groups, self.n_splits)), Warning)


0.03353378010399599
0.02266811516995282
0.02667403430072035
0.01395320028874869
0.030367617948803827


### Including things we know likes are dependent on 

In [31]:
X = data_new[['Followings', 'Followers', 'MediaCount','Selfie', 'BodySnap', 'Marketing', 'ProductOnly', 'NonFashion',
       'Face', 'Logo', 'BrandLogo', 'Smile', 'Outdoor', 'NumberOfPeople',
       'NumberOfFashionProduct', 'Anger', 'Contempt', 'Disgust', 'Fear',
       'Happiness', 'Neutral', 'Sadness', 'Surprise', 'Designer',
       'High street', 'Mega couture', 'Small couture']]
y = data_new['Likes']

In [32]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.33, random_state=42)

In [69]:
def standardize(train, test):


    mean = np.mean(train, axis=0)
    std = np.std(train, axis=0)+0.000001

    X_train = (train - mean) / std
    X_test = (test - mean) /std
    return X_train, X_test

In [70]:
X_train, X_test = standardize(X_train, X_test)

Let's use a Lasso Regression model to regress based on the likes that we receive. I want to use lasso because lasso will send the unimportant variables to zero allowing us to see which features are important. 

In [71]:
lasso = Lasso(alpha = 0.1, max_iter = 10000).fit(X_train, y_train)

In [72]:
predictions = lasso.predict((X_test))
r2_score(y_test, predictions)

0.2984377062815089

Maybe ridge regression will work better?

In [73]:
ridge = Ridge(alpha = 0.1).fit(X_train, y_train)

In [74]:
predictions = ridge.predict((X_test))
r2_score(y_test, predictions)

0.2984484171726812

So ridge regression is really no different...

### Try a Deep Learning Model

In [45]:
def our_model():
    model = Sequential()

    model.add(Dense(units=64, activation='relu', input_dim = 24))
    model.add(Dense(units=64, activation='relu'))
    model.add(Dense(units=64, activation='relu'))
    model.add(Dense(units=1))
    model.compile(loss='mean_squared_error', optimizer='adam',metrics=['accuracy'])
    return model

In [46]:
estimator = KerasRegressor(build_fn=our_model, epochs=100, batch_size=5, verbose=1)

In [47]:
from sklearn.model_selection import KFold, cross_val_score

kfold = KFold(n_splits=3)
results = cross_val_score(estimator, X_train, y_train, cv=kfold)
print("Baseline: %.2f (%.2f) MSE" % (results.mean(), results.std()))

Epoch 1/100
11055/11055 [==============================] - 2s 184us/step - loss: 3641.0656 - accuracy: 0.0014
Epoch 2/100
11055/11055 [==============================] - 2s 157us/step - loss: 3565.2703 - accuracy: 4.5228e-04
Epoch 3/100
 3490/11055 [========>.....................] - ETA: 1s - loss: 3531.1837 - accuracy: 2.8653e-04

KeyboardInterrupt: 

In [48]:
estimator.fit(X_train, y_train)

Epoch 1/100
16583/16583 [==============================] - 3s 175us/step - loss: 3519.1527 - accuracy: 0.0019
Epoch 2/100
16583/16583 [==============================] - 3s 169us/step - loss: 3449.8860 - accuracy: 0.0011
Epoch 3/100
16583/16583 [==============================] - 3s 157us/step - loss: 3425.9882 - accuracy: 8.4424e-04
Epoch 4/100
16583/16583 [==============================] - 3s 159us/step - loss: 3406.3561 - accuracy: 9.6484e-04
Epoch 5/100
16583/16583 [==============================] - 3s 160us/step - loss: 3383.3030 - accuracy: 9.6484e-04
Epoch 6/100
16583/16583 [==============================] - 3s 158us/step - loss: 3376.5649 - accuracy: 0.0010
Epoch 7/100
16583/16583 [==============================] - 3s 156us/step - loss: 3364.3806 - accuracy: 0.0011
Epoch 8/100
16583/16583 [==============================] - 3s 155us/step - loss: 3350.8871 - accuracy: 6.6333e-04
Epoch 9/100
16583/16583 [==============================] - 3s 161us/step - loss: 3351.8270 - accuracy: 7

In [49]:
preds = estimator.predict(X_test)

8169/8169 [==============================] - 0s 33us/step


In [50]:
r2_score(y_test, preds)

-0.00927594237649898

Maybe we need to investigate how the input space was normalized. Maybe the categorical variables that are represented by 1's and 0's at the end of the dataframe need to be normalized?

After log normalization of the likes we have worse results on this model.

Average loss value for Keras deep learning model based on just numeric values and predicting likes is 154K.

Maybe try using likes / followers?

# Likes to Follower Ratio

First check if there are any null likes or follower entries

In [25]:
data_new['Likes'].isnull().any() 

False

In [26]:
data_new['Followers'].isnull().any()

False

In [27]:
data_ver2 = pd.concat([data, brand_category], sort = False, axis = 1)
data_ver2['LikesToFollowers'] = data_ver2['Likes'] / data_ver2['Followers']
data_ver2.head()

Followings  Followers  MediaCount  \
UserId                                                              
1171579752855683619_212070047        518.0     9840.0         541   
1171594777274371222_176762322       7333.0     2300.0         272   
1171407552643586413_581125501        131.0      605.0         106   
1171407552643586413_581125501        131.0      605.0         106   
1171508187966229230_2797323089       132.0      456.0          56   

                                  BrandName BrandCategory  \
UserId                                                      
1171579752855683619_212070047   abercrombie   High street   
1171594777274371222_176762322   abercrombie   High street   
1171407552643586413_581125501   abercrombie   High street   
1171407552643586413_581125501   abercrombie   High street   
1171508187966229230_2797323089  abercrombie   High street   

                                                                         Hashtags  \
UserId                                                                              
1171579752855683619_212070047   beautiful, summer, fashion, love, cute, food, ...   
1171594777274371222_176762322   teen, model, brunette, selfie, hollister, snap...   
1171407552643586413_581125501   alexandani, cute, llbean, beanboots, anthropol...   
1171407552643586413_581125501   alexandani, cute, llbean, beanboots, anthropol...   
1171508187966229230_2797323089  abercrombieandfitch, shopping, love, hollister...   

                                                                          Caption  \
UserId                                                                              
1171579752855683619_212070047     We were born to be REAL, not to be perfect. ...   
1171594777274371222_176762322     Gotta run, but first, let me take selfie. Me...   
1171407552643586413_581125501                         dress to impress\U0001f457    
1171407552643586413_581125501                         dress to impress\U0001f457    
1171508187966229230_2797323089    #love #shopping #shoppen #hollister #abercro...   

                                                                           ImgURL  \
UserId                                                                              
1171579752855683619_212070047   https://scontent.cdninstagram.com/t51.2885-15/...   
1171594777274371222_176762322   https://scontent.cdninstagram.com/t51.2885-15/...   
1171407552643586413_581125501   https://scontent.cdninstagram.com/t51.2885-15/...   
1171407552643586413_581125501   https://scontent.cdninstagram.com/t51.2885-15/...   
1171508187966229230_2797323089  https://scontent.cdninstagram.com/t51.2885-15/...   

                                Likes  Comments   CreationTime  \
UserId                                                           
1171579752855683619_212070047      97          0    1453883211   
1171594777274371222_176762322      94          0    1453885002   
1171407552643586413_581125501      91          3    1453862683   
1171407552643586413_581125501      94          3    1453862683   
1171508187966229230_2797323089      9          1    1453874680   

                                                                    Link  \
UserId                                                                     
1171579752855683619_212070047   https://www.instagram.com/p/BBCSZPzAMIj/   
1171594777274371222_176762322   https://www.instagram.com/p/BBCVz4YJEyW/   
1171407552643586413_581125501   https://www.instagram.com/p/BBBrPZ2h9lt/   
1171407552643586413_581125501   https://www.instagram.com/p/BBBrPZ2h9lt/   
1171508187966229230_2797323089  https://www.instagram.com/p/BBCCH1zMEru/   

                                  Selfie  BodySnap  Marketing  ProductOnly  \
UserId                                                                       
1171579752855683619_212070047   0.000885  0.336438   0.741220     0.009454   
1171594777274371222_176762322   0.003912  0.995503   0.001728     0.002125   
1171407552643586413_5811

In [28]:
data_ver2.replace([np.inf, -np.inf], np.nan, inplace = True);

In [29]:
data_test = data_ver2[data_ver2['LikesToFollowers'].isnull()]
data_test.head()

Followings  Followers  MediaCount  \
UserId                                                              
1171240019544303946_23457617           0.0        0.0           1   
1171142669094830425_252313279         36.0        0.0           0   
1171278619918133139_23638768           0.0        0.0           0   
1170503373243750859_2795691477         0.0        0.0           1   
1170007750938083500_2120361935        39.0        0.0           0   

                                    BrandName BrandCategory  \
UserId                                                        
1171240019544303946_23457617      acnestudios      Designer   
1171142669094830425_252313279     acnestudios      Designer   
1171278619918133139_23638768      acnestudios      Designer   
1170503373243750859_2795691477    acnestudios      Designer   
1170007750938083500_2120361935  alexanderwang      Designer   

                                                                         Hashtags  \
UserId                                                                              
1171240019544303946_23457617    shop, anndemeulemeester, style, acnestudios, t...   
1171142669094830425_252313279                            acnestudios, collegeface   
1171278619918133139_23638768                                                  NaN   
1170503373243750859_2795691477  sorpreseinmacchina, acnestudios, withstyle, ga...   
1170007750938083500_2120361935  jimmychoo, slimthickwithyourcuteass, alexander...   

                                                                          Caption  \
UserId                                                                              
1171240019544303946_23457617      Black? Or dark black... @00000 #totokaelo #l...   
1171142669094830425_252313279     #collegeface in college\U0001f610 #acnestudios    
1171278619918133139_23638768                1.26 | \u2603  Cc: \U0001f4f7 @00000    
1170503373243750859_2795691477    #jackolsen #jauria #withstyle #migraineprobl...   
1170007750938083500_2120361935    u\"\"I swear she's so perfect she makes me s...   

                                                                           ImgURL  \
UserId                                                                              
1171240019544303946_23457617    https://scontent.cdninstagram.com/t51.2885-15/...   
1171142669094830425_252313279   https://scontent.cdninstagram.com/t51.2885-15/...   
1171278619918133139_23638768    https://scontent.cdninstagram.com/t51.2885-15/...   
1170503373243750859_2795691477  https://scontent.cdninstagram.com/t51.2885-15/...   
1170007750938083500_2120361935  https://scontent.cdninstagram.com/t51.2885-15/...   

                                Likes  Comments   CreationTime  \
UserId                                                           
1171240019544303946_23457617       37          0    1453842712   
1171142669094830425_252313279      18          1    1453831107   
1171278619918133139_23638768      136          9    1453847313   
1170503373243750859_2795691477     10          0    1453754897   
1170007750938083500_2120361935     74          0    1453695814   

                                                                    Link  \
UserId                                                                     
1171240019544303946_23457617    https://www.instagram.com/p/BBBFJefqOlK/   
1171142669094830425_252313279   https://www.instagram.com/p/BBAvA10s0FZ/   
1171278619918133139_23638768    https://www.instagram.com/p/BBBN7L5OueT/   
1170503373243750859_2795691477  https://www.instagram.com/p/BA-dp3KlxnL/   
1170007750938083500_2120361935  https://www.instagram.com/p/BA8s9m7HkCs/   

                                  Selfie  BodySnap  Marketing  ProductOnly  \
UserId                                                                       
1171240019544303946_23457617    0.000158  0.001617   0.007038     0.997941   
1171142669094830425_252313279   0.870760  0.200230   0.063385     0.003886   
1171278619

Drop rows where LikesToFollowers is NAN. Likes and Followers count at both 0.

In [30]:
data_ver2 = data_ver2.dropna()

In [31]:
data_ver2['LikesToFollowers'].isnull().any()

False

In [32]:
X = data_ver2[['Selfie', 'BodySnap', 'Marketing', 'ProductOnly', 'NonFashion',
       'Face', 'Logo', 'BrandLogo', 'Smile', 'Outdoor', 'NumberOfPeople',
       'NumberOfFashionProduct', 'Anger', 'Contempt', 'Disgust', 'Fear',
       'Happiness', 'Neutral', 'Sadness', 'Surprise', 'Designer',
       'High street', 'Mega couture', 'Small couture']]
y = data_ver2['LikesToFollowers']

In [101]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=42)

In [33]:
def standardize(train, test):


    mean = np.mean(train, axis=0)
    std = np.std(train, axis=0)+0.000001

    X_train = (train - mean) / std
    X_test = (test - mean) /std
    return X_train, X_test

In [103]:
X_train, X_test = standardize(X_train, X_test)

In [104]:
epsilon = 1e-200
y_train = np.log(y_train+epsilon)
y_test = np.log(y_test+epsilon)

In [105]:
import keras
keras.backend.floatx()
keras.backend.set_floatx('float64')
keras.backend.floatx()

'float64'

In [34]:
def our_model_ver2():
    model = Sequential()
    model.add(Dense(units=128, input_dim = 24))
    model.add(LeakyReLU(alpha=0.1))
    model.add(Dropout(0.25))

    model.add(Dense(units=256))
    model.add(LeakyReLU(alpha=0.1))
    model.add(Dropout(0.25))
    
    model.add(Dense(units=64, activation = 'tanh'))
    
    model.add(Dense(units=1))
    #opt = Adadelta(clipnorm=0.3)
    model.compile(loss='mean_squared_error', optimizer='adam',metrics=['accuracy'])
    return model

estimator = KerasRegressor(build_fn=our_model_ver2, epochs=100, batch_size=500, verbose=1)

# kfold = KFold(n_splits=5)
# results = cross_val_score(estimator, X_train, y_train, cv=kfold)
# print("Baseline: %.2f (%.2f) MSE" % (results.mean(), results.std()))

In [107]:
estimator.fit(X_train, y_train)

Epoch 1/100
22088/22088 [==============================] - 0s 21us/step - loss: 3501.9655 - accuracy: 0.0000e+00
Epoch 2/100
22088/22088 [==============================] - 0s 12us/step - loss: 3456.3082 - accuracy: 0.0000e+00
Epoch 3/100
22088/22088 [==============================] - 0s 11us/step - loss: 3438.2888 - accuracy: 0.0000e+00
Epoch 4/100
22088/22088 [==============================] - 0s 12us/step - loss: 3420.8879 - accuracy: 0.0000e+00
Epoch 5/100
22088/22088 [==============================] - 0s 12us/step - loss: 3410.5549 - accuracy: 0.0000e+00
Epoch 6/100
22088/22088 [==============================] - 0s 12us/step - loss: 3398.5681 - accuracy: 0.0000e+00
Epoch 7/100
22088/22088 [==============================] - 0s 11us/step - loss: 3391.0990 - accuracy: 0.0000e+00
Epoch 8/100
22088/22088 [==============================] - 0s 12us/step - loss: 3387.6629 - accuracy: 0.0000e+00
Epoch 9/100
22088/22088 [==============================] - 0s 12us/step - loss: 3382.5712 - accu

In [108]:
preds = estimator.predict(X_test)

2455/2455 [==============================] - 0s 16us/step


In [109]:
r2_score(y_test, preds)

0.08561252991158974

When using log normalization of likes/followers we got 0.036 and when we did log normalization of just likes we got 0.035. Additionally, reducing the amount of test data and the batch size improves the results further to 0.0856.

### Try stratified crossvalidation

In [35]:
cv = StratifiedKFold(5)
for train, test in cv.split(X, y):
    X_train, X_test = standardize(X[train], X[test])
    epsilon = 1e-200
    y_train = np.log(y[train] + epsilon)
    y_test = np.log(y[test] + epsilon)
    
    estimator.fit(X_train, y_train)
    predictions = estimator.predict((X_test))
    print(r2_score(y_test, predictions))

ValueError: Supported target types are: ('binary', 'multiclass'). Got 'continuous' instead.

 okay so we can't use this for likes to followers